arangod/Aql/SubqueryEndExecutor.cpp

```diff
@@ -0,0 +1,182 @@
+          //       into the SubqueryEndNode
+          TRI_ASSERT(_accumulator.isOpenArray());
+          AqlValue value = inputRow.getValue(0);
+          value.toVelocyPack(_infos.getTrxPtr(), _accumulator, false);
+        }
+
+        // We have received DONE on data rows, so now
+        // we have to read a relevant shadow row
+        if (state == ExecutionState::DONE) {
+          _accumulator.close();
+          TRI_ASSERT(_accumulator.isClosed());
+          _state = RELEVANT_SHADOW_ROW_PENDING;
+        }
+        break;
+      }
+      case RELEVANT_SHADOW_ROW_PENDING: {
+        std::tie(state, shadowRow) = _fetcher.fetchShadowRow();
+        if (state == ExecutionState::WAITING) {
+          TRI_ASSERT(!shadowRow.isInitialized());
+          return {ExecutionState::WAITING, std::move(stats)};
+        }
+        TRI_ASSERT(state == ExecutionState::DONE || state == ExecutionState::HASMORE);
+        TRI_ASSERT(shadowRow.isInitialized() == true);
+        TRI_ASSERT(shadowRow.isRelevant() == true);
+
+        // Here we have all data *and* the relevant shadow row,
+        // so we can now submit
+        AqlValue resultDocVec{_accumulator.slice()};
+        AqlValueGuard guard{resultDocVec, true};
+
+        // Responsibility is handed over
+        output.consumeShadowRow(_infos.getOutputRegister(), shadowRow, guard);
+        TRI_ASSERT(output.produced());
+        output.advanceRow();
+
+        // Reset the accumulator in case we read more data
+        resetAccumulator();
+
+        if (state ==  ExecutionState::DONE) {
+          return {ExecutionState::DONE, std::move(stats)};
+        } else {
+          _state = FORWARD_IRRELEVANT_SHADOW_ROWS;
+        }
+        break;
+      }
+      case FORWARD_IRRELEVANT_SHADOW_ROWS: {
+        // Forward irrelevant shadow rows (and only those)
+        std::tie(state, shadowRow) = _fetcher.fetchShadowRow();
+
+        if (state == ExecutionState::WAITING) {
+          return {ExecutionState::WAITING, std::move(stats)};
+        }
+
+        if (shadowRow.isInitialized()) {
+          // We got a shadow row, it must be irrelevant,
+          // because to get another relevant shadowRow we must
+          // first call fetchRow again
+          TRI_ASSERT(shadowRow.isRelevant() == false);
+          output.decreaseShadowRowDepth(shadowRow);
+          output.advanceRow();
+        } else {
+          // We did not get another shadowRow; either we
+          // are DONE or we are getting another relevant
+          // shadow row, but only after we called fetchRow
+          // again
+          if (state == ExecutionState::HASMORE) {
+            _state = ACCUMULATE;
+          }
+        }
+
+        if (state == ExecutionState::DONE) {
+          return {ExecutionState::DONE, std::move(stats)};
+        }
+        break;
+      }
+
+      default: {
+        TRI_ASSERT(false);
+        break;
+      }
+    }
+  }
+  // We should *only* fall through here if output.isFull() is true.
```

Comment:
NOTE:
I like this, from my point of view it was straight forward to understand now what this executor is doing :+1: